### DISCUSSION TITLE
HMM for one-dimensional data with negative step lengths

### DISCUSSION THREAD

**Discussion by mischaschultz** [bmcclintock/momentuHMM#106] [Q&A] [answered]
I am interested in using momentuHMM to model one-dimensional movement patterns. For now, I am starting off with just a simple HMM. I have do not have angles in my data, just step lengths. My data is formatted as a change in depth from time t to time t +1 (both negative and positive values) in 5 minute intervals. I want to model the movement path with three known states: ascending (+), descending (-), and constant (no change, 0). 

Below is an example of the output from momentuhmm:prepData

##prep the data 
> HMM_prep <- momentuHMM::prepData(data=as.data.frame(sharkData), coordNames=NULL)
> head(HMM_prep)
       ID                time   step     state  temp
1 Animal1 2021-04-17 15:44:00  -0.25    diving 17.59
2 Animal1 2021-04-17 15:49:00  21.25 ascending 17.85
3 Animal1 2021-04-17 15:54:00   5.00 ascending 16.83
4 Animal1 2021-04-17 15:59:00 -26.25    diving 15.05
5 Animal1 2021-04-17 16:04:00   0.00  constant 17.59
6 Animal1 2021-04-17 16:09:00   0.00  constant 17.59
> summary(HMM_prep, dataNames = "step")
HMM data for 1 individual:

Animal1 -- 7160 observations

Data summaries:

      step         
 Min.   :-2.1e+02  
 1st Qu.:-2.0e+00  
 Median : 0.0e+00  
 Mean   :-3.0e-05  
 3rd Qu.: 2.0e+00  
 Max.   : 2.5e+02  
> nbStates = 3                                      #number of states
> stateNames = c("Diving", "Constant", "Ascending") #state names
>

#parameters and probability distributions
stepPar = **This is where I am getting messed up choosing the parameters because I have negative values
dist = list(step="norm")
stationary = TRUE

fit <- momentuHMM::fitHMM(data = HMM_prep,
                          nbStates = nbStates,
                          dist = dist,
                          Par0 = list(step=stepPar)) 

I am fitting step length as a "norm" distribution. I am curious how to run this model when there are only step lengths and some contain negative values. Are you able to select negative values for initial starting parameters for mean and sd? I get this error message when trying to set initial parameters for each state: Check the parameter bounds for step (the initial parameters should be strictly between the bounds of their parameter space)

Let me know if I need to provide more information.

**Comment by bmcclintock** [comment, marked as answer]
I suppose you could try and model these using a normal distribution, in which case `mean` can be any real value (positive or negative) and `sd` must be positive. As long as these conditions are met, you should not be getting an error message; thus, in addition to checking that `sd` is positive, you might want to check the order of your `stepPar` initial values, which should be `c(mean_1, mean_2, mean_3, sd_1, sd_2, sd_3)`.

However, one potential problem with the normal distribution is that this would not necessarily or effectively forbid "Diving" from being positive, "Constant" from being negative or positive (unless of course `mean_2` is fixed to zero and `sd_2` is fixed to an extremely small positive value), and "Ascending" from being negative. This may not be a problem if the "Diving" step values are sufficiently negative (i.e. most/all much less than 0) and the "Ascending" values are sufficiently positive (i.e. most/all much greater than 0). If not, then maybe a little bit of wiggle room would be OK (e.g. if the occasional positive step during "Diving" and the occasional negative step during "Ascending" is acceptable). Another issue is that a normal distribution may not be able to adequately capture a potential left-skewed tail for "Diving" and a potential right-skewed tail for "Ascending", in which case the `t` distribution might be better.

That said, if all you really care about is "Diving", "Constant", and "Ascending" (with little concern for the actual changes in depth), then you could instead assign each vertical step to one of these three categories and fit these categorical data using the `cat` distribution.

Assuming normality is appropriate, here is some code to simulate depths as a random walk with three states and then fit a 3-state model for the vertical steps. Hopefully you can play around with this and get a better sense for how appropriate this approach may be for your data.
```
library(momentuHMM)
library(tidyverse)
set.seed(1,kind="Mersenne-Twister",normal.kind="Inversion")
simdat <- simData(nbAnimals=2,obsPerAnimal=1000,nbStates=3,dist=list(mu="rw_mvnorm2"),
                  DM=list(mu=matrix(c("mu.x_tm1",1,0,0,0,0,0,
                                      "mu.x_tm1",0,0,0,0,0,0,
                                      "mu.x_tm1",0,1,0,0,0,0,
                                      "mu.y_tm1",1,0,0,0,0,0,
                                      "mu.y_tm1",0,0,0,0,0,0,
                                      "mu.y_tm1",0,1,0,0,0,0,
                                               0,0,0,1,0,0,0,
                                               0,0,0,0,1,0,0,
                                               0,0,0,0,0,1,0,
                                               0,0,0,0,0,0,1,
                                               0,0,0,0,0,0,1,
                                               0,0,0,0,0,0,1,
                                               0,0,0,1,0,0,0,
                                               0,0,0,0,1,0,0,
                                               0,0,0,0,0,1,0),15,7,byrow=TRUE)),
                  Par=list(mu=c(1,-20,20,log(40),log(.25),log(40),0)),beta=matrix(c(-3.5,-2.5,-1,-1,-2.5,-3.5),1,6),states=TRUE,stateNames=c("Diving","Constant","Ascending"),mvnCoords = "mu")

simdat <- simdat %>% rename(depth=mu.x) %>% select(ID,depth,states)
plot(simdat,dataNames="depth",ask=FALSE)
simdat$step <- unlist(mapply(function(x) c(diff(simdat$depth[which(simdat$ID==x)]),NA),unique(simdat$ID),SIMPLIFY = FALSE))
fit <- fitHMM(simdat,nbStates=3,dist=list(step="norm"),Par0=list(step=c(-20,0,20,10,0.1,10)),stateNames=c("Diving","Constant","Ascending"))
fit
plot(fit,ask=FALSE)
sum(viterbi(fit)==simdat$states)/nrow(simdat)
plotPR(fit)
```
![image](https://user-images.githubusercontent.com/11667599/139155704-80ed728d-424c-4092-87be-6dc96a8b0e63.png)
